### PR TITLE
Add EnumJsonAdapter to adapters module.

### PR DESCRIPTION
**Pull request by NightlyNexus**
for consideration. dunno what to call this type.

### CODE REVIEW

File: adapters/src/main/java/com/squareup/moshi/adapters/ConfigurableEnumJsonAdapter.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
+package com.squareup.moshi.adapters;
```
Comment (swankjesse) [inline adapters/src/main/java/com/squareup/moshi/adapters/ConfigurableEnumJsonAdapter.java]: nit: copyright header

File: adapters/src/main/java/com/squareup/moshi/adapters/ConfigurableEnumJsonAdapter.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+import com.squareup.moshi.JsonWriter;
+import java.io.IOException;
+import java.util.Arrays;
+import javax.annotation.Nullable;
+
+/**
+ * A JsonAdapter for enums that allows having a fallback enum value when a deserialized string does
+ * not match any enum value.
+ */
+public final class ConfigurableEnumJsonAdapter<T extends Enum<T>> extends JsonAdapter<T> {
```
Comment (swankjesse) [inline adapters/src/main/java/com/squareup/moshi/adapters/ConfigurableEnumJsonAdapter.java]: I wonder if this should be just called EnumJsonAdapter ? The other one isn’t a public type so I don’t think it matters

File: adapters/src/main/java/com/squareup/moshi/adapters/ConfigurableEnumJsonAdapter.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  final String[] nameStrings;
+  final T[] constants;
+  final JsonReader.Options options;
+  final @Nullable T fallbackValue;
+
+  public static <T extends Enum<T>> ConfigurableEnumJsonAdapter<T> create(Class<T> enumType) {
+    return new ConfigurableEnumJsonAdapter<>(enumType, null);
+  }
+
+  public ConfigurableEnumJsonAdapter<T> withFallbackValue(T fallbackValue) {
```
Comment (swankjesse) [inline adapters/src/main/java/com/squareup/moshi/adapters/ConfigurableEnumJsonAdapter.java]: Should document when we fallback? Like only when an enum is unknown. (But not absent or null!)

Maybe should be `withUnknownValue()` ?

File: adapters/src/main/java/com/squareup/moshi/adapters/ConfigurableEnumJsonAdapter.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+  ConfigurableEnumJsonAdapter(Class<T> enumType, @Nullable T fallbackValue) {
+    this.enumType = enumType;
+    this.fallbackValue = fallbackValue;
+    try {
+      constants = enumType.getEnumConstants();
+      nameStrings = new String[constants.length];
+      for (int i = 0; i < constants.length; i++) {
+        T constant = constants[i];
+        Json annotation = enumType.getField(constant.name()).getAnnotation(Json.class);
```
Comment (swankjesse) [inline adapters/src/main/java/com/squareup/moshi/adapters/ConfigurableEnumJsonAdapter.java]: this is great

File: adapters/src/main/java/com/squareup/moshi/adapters/ConfigurableEnumJsonAdapter.java (hunk 3 above)
Comment (swankjesse) [reply adapters/src/main/java/com/squareup/moshi/adapters/ConfigurableEnumJsonAdapter.java]: Might be extra worth a warning to say that this adapter is case-sensitive and that the fallback will be used even on just a case mismatch

File: adapters/src/test/java/com/squareup/moshi/adapters/ConfigurableEnumJsonAdapterTest.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
+package com.squareup.moshi.adapters;
```
Comment (swankjesse) [inline adapters/src/test/java/com/squareup/moshi/adapters/ConfigurableEnumJsonAdapterTest.java]: nit: copyright header

File: adapters/src/test/java/com/squareup/moshi/adapters/ConfigurableEnumJsonAdapterTest.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    try {
+      adapter.fromJson("\"SPOCK\"");
+      fail();
+    } catch (JsonDataException expected) {
+      assertThat(expected).hasMessage(
+          "Expected one of [ROCK, PAPER, scr] but was SPOCK at path $");
+    }
+  }
+
+  @Test public void withFallbackValue() throws Exception {
```
Comment (swankjesse) [inline adapters/src/test/java/com/squareup/moshi/adapters/ConfigurableEnumJsonAdapterTest.java]: nice

File: adapters/src/main/java/com/squareup/moshi/adapters/ConfigurableEnumJsonAdapter.java (hunk 2 above)
Comment (NightlyNexus) [reply adapters/src/main/java/com/squareup/moshi/adapters/ConfigurableEnumJsonAdapter.java]: yes, totally. good call.

File: adapters/src/main/java/com/squareup/moshi/adapters/ConfigurableEnumJsonAdapter.java (hunk 3 above)
Comment (NightlyNexus) [reply adapters/src/main/java/com/squareup/moshi/adapters/ConfigurableEnumJsonAdapter.java]: oops, i missed putting in a doc. will update.

`withUnknownValue` reads strange to me. it sounds like the value passed in is unknown, when it is really the substitute for unknown values.

Comment (swankjesse) [reply adapters/src/main/java/com/squareup/moshi/adapters/ConfigurableEnumJsonAdapter.java]: Maybe withUnknownFallback() ?

Comment (NightlyNexus) [reply adapters/src/main/java/com/squareup/moshi/adapters/ConfigurableEnumJsonAdapter.java]: sounds good.

File: adapters/src/main/java/com/squareup/moshi/adapters/EnumJsonAdapter.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    if (fallbackValue == null) {
+      String name = reader.nextString();
+      throw new JsonDataException("Expected one of "
+          + Arrays.asList(nameStrings) + " but was " + name + " at path " + path);
+    }
+    if (reader.peek() != JsonReader.Token.STRING) {
+      throw new JsonDataException(
+          "Expected a string but was " + reader.peek() + " at path " + path);
+    }
+    reader.skipValue();
```
Comment (NightlyNexus) [inline adapters/src/main/java/com/squareup/moshi/adapters/EnumJsonAdapter.java]: changed to skipValue just now.

**Comment by edenman** [comment]
Is there a reason this doesn't allow `null` as the fallback value?

**Comment by edenman** [comment]
/shrug I find it cleaner to model as `MyEnumType?`...seems weird to prescribe an approach: why not just default to the "return null if not a known value" behavior and instead use a fallback if one was provided?  I guess this might be why moshi didn't ship with one of these in the first place, everybody wants something slightly different :)

**Comment by hzsweers** [comment]
you can exhaustively switch on unknown, I find that to be a nicer pattern

**Comment by JakeWharton** [comment]
Not allowing null is weird. What if you don't control the enum and thus cannot add an `UNKOWN`?

**Comment by shanyuliang** [comment]
Agree with @edenman , in my working project we do have 10+ enum classes. For centain enum class we sure can provide our own fallback value. But would be cool if we have a general "enum-fallback-to-null" solution for all other enums.

**Comment by NightlyNexus** [comment]
The factory that makes enum adapters that always fall back to the universal null value is tempting.

As for this adapter, I'm really not opinionated about adding null as a possible fallback value. PR welcome, or I'll get to it at some point. Just needs a flag to track if a fallback has been set for the adapter.

**Comment by swankjesse** [comment]
Yeah I think a null fallback is handy.

**Comment by NightlyNexus** [comment]
#640

**Pull request merged**